### DISCUSSION TITLE
Replace inplace?

### DISCUSSION THREAD

**Discussion by jmigual** [BurntSushi/ripgrep#2568] [General] [answered]
So, I've discovered the `-r` switch to replace some text. However, I have about 50k files where I would like to replace a string. I was expecting the command `rg "out/gen" -r "data/gen"` to replace all the "out/gen" string inside all my files with "data/gen" but I later realized that this does not work.

I ended up using VS Code to perform that task which was very slow. Would it be possible to add a switch to do the replacement inplace instead of just printing it?

**Comment by s-p-turner** [comment, marked as answer]
The documentation for the _-r_ flag states: "Replace every match with the text given when printing results. Neither this flag nor any other ripgrep flag will modify your files."

I doubt the authors would want to modify ripgrep to do in-place editing of files. And neither should they in my opinion, as it's a file search utility and not a file editing utility. I would suggest either writing a script and/or using the _find_ utility to iterate over your files and use _sed -i_ to replace your text instead.

**Reply by BurntSushi** [reply]
See also https://github.com/BurntSushi/ripgrep/issues/74 and https://github.com/BurntSushi/ripgrep/blob/master/FAQ.md#search-and-replace.

**Reply by aagit** [reply]
Hello,

> See also #74 and https://github.com/BurntSushi/ripgrep/blob/master/FAQ.md#search-and-replace.

I just noticed this link on this old thread. Replace-inplace/search-and-replace can also be achieved with [ripgrep-edit](https://gitlab.com/aarcange/ripgrep-edit).

`ripgrep-edit` is particularly valuable for complex refactorings where LLM assistance is needed (optionally aided by dynamically generated GBNF grammar), but it can also be used for straightforward text replacements.

I wonder if you would consider adding `ripgrep-edit` to the ripgrep FAQ as an alternative to `sed` for more complex razor coding workflows with LLM assistance. The tool is specifically designed to addresses the limitations of traditional approaches where the grep tool is entirely offloaded to the agent.

**Comment by s-p-turner** [comment]
It's probably worth mentioning that _sed -i ..._ will destroy symbolic links if you do an in-place replacement of text in a file that happens to be a symbolic link. The way to get around this behaviour is to use the _--follow-symlinks_ argument i.e. _sed --follow-symlinks -i..._ I don't know whether this might be worth adding to the ripgrep FAQ? I will of course understand if you don't feel this is within the remit of ripgrep. But you do mention sed as a possible solution to replacing text in files in the FAQ already ;-)

**Reply by BurntSushi** [reply]
I don't think so. The point of the FAQ item is to give folks some pointers. But I really do not want to get in the business of maintaining a complete guide to in-place replacement.

**Reply by BurntSushi** [reply]
(I have also never used `--follow-symlinks` with `sed -i` myself and it hasn't bitten me in a couple decades. Doesn't mean it isn't valid, but might point to it being a niche issue.)

**Reply by s-p-turner** [reply]
After 40 years of developing on Unix/Linux, it happened to bite me for the first ever time last week. Which is why I thought I'd mention it as the wound hasn't yet healed. But as I stated earlier, I completely understand your decision to not add this to the ripgrep FAQ.

**Comment by gokayburuc** [comment]
Write some script using `rg` and `sed` altogether like given below: 

```bash 
for file in $(rg --files)
do
  new=$(echo "$file" | sed -e 's/bljuz_muzika_-_//' -e 's/_(z3.fm)//' -e 's/_-_/_/g')
  mv $file $new
  echo "Renamed $file --> $new"
done

```

I tested this small script with my mp3 files about 120 files. Working properly. `sed` gives you to make CRUD operations using `regex` 

You can rename 50k files with this small script.

That's all.

**Reply by lespea** [reply]
You can also check out https://github.com/chmln/sd
